 **Task description**

The goal is to analyze the database of the purchased service for reading books by subscription in order to further formulate a value proposition for a new product.

 **Goal:**

* get acquainted with the data stored in the database,
parse query results.

The data is arranged in 5 tables containing information about books, publishers, authors, user ratings and book reviews.

In [2]:
# import libraries
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

# set configuration for database connection
db_config = {'user': 'praktikum_student', # user name
             'pwd': 'Sdf4$2;d-d30pp',     # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,                # connection port
             'db': 'data-analyst-final-project-db'}  # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

* **Study the tables (print the first rows)**

In [3]:
tables_list = ['books', 'authors', 'publishers', 'ratings', 'reviews']

In [11]:
for t in tables_list:
    print('Table {} head'.format(t))
    display(pd.io.sql.read_sql('''SELECT * FROM {} LIMIT 5'''.format(t), con = engine))
    print()
    print('Data types')
    display(pd.io.sql.read_sql('''  SELECT column_name, data_type
                                    FROM information_schema.columns
                                    WHERE table_name = '{}';'''.format(t), con = engine))
    print()
    print('Missing values')
    display(pd.io.sql.read_sql(''' SELECT column_name, 
                                   SUM(CASE WHEN column_name IS NULL THEN 1 ELSE 0 END) / SUM(1.0) AS null_ratio
                                   FROM information_schema.columns 
                                   WHERE table_name = '{}'
                                   GROUP BY column_name;'''.format(t), con = engine))
    print()

Table books head


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


Data types


column_name data_type
0           book_id   integer
1         author_id   integer
2             title      text
3         num_pages   integer
4  publication_date      date
5      publisher_id   integer


Missing values


column_name  null_ratio
0         author_id         0.0
1           book_id         0.0
2         num_pages         0.0
3  publication_date         0.0
4      publisher_id         0.0
5             title         0.0


Table authors head


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


Data types


column_name data_type
0   author_id   integer
1      author      text


Missing values


column_name  null_ratio
0      author         0.0
1   author_id         0.0


Table publishers head


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


Data types


column_name data_type
0  publisher_id   integer
1     publisher      text


Missing values


column_name  null_ratio
0     publisher         0.0
1  publisher_id         0.0


Table ratings head


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


Data types


column_name data_type
0   rating_id   integer
1     book_id   integer
2    username      text
3      rating   integer


Missing values


column_name  null_ratio
0     book_id         0.0
1      rating         0.0
2   rating_id         0.0
3    username         0.0


Table reviews head


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...


Data types


column_name data_type
0   review_id   integer
1     book_id   integer
2    username      text
3        text      text


Missing values


column_name  null_ratio
0     book_id         0.0
1   review_id         0.0
2        text         0.0
3    username         0.0

**Find the number of books released after January 1, 2000.**

In [4]:
query1 = """SELECT COUNT(book_id)
            FROM books
            WHERE publication_date > '2000-01-01'"""
pd.io.sql.read_sql(query1, con = engine)

count
0    819

There were 819 books published after January 1, 2000.

**Find the number of user reviews and the average rating for each book.**

In [9]:
query2 = ''' SELECT title, cnt_review , avg_raiting
            FROM books 
            LEFT JOIN (SELECT book_id, COUNT(review_id) AS cnt_review 
                        FROM reviews
                        GROUP BY book_id) AS sub
            ON books.book_id = sub.book_id
            LEFT JOIN (SELECT book_id, AVG(rating) AS avg_raiting
                        FROM ratings 
                        GROUP BY book_id) AS sub_2
            ON books.book_id = sub_2.book_id
        '''
pd.io.sql.read_sql(query2, con = engine)

title  cnt_review  \
0            The Body in the Library (Miss Marple  #3)         2.0   
1                                            Galápagos         2.0   
2                             A Tree Grows in Brooklyn         5.0   
3    Undaunted Courage: The Pioneering First Missio...         2.0   
4                                          The Prophet         4.0   
..                                                 ...         ...   
995                                Alice in Wonderland         4.0   
996          A Woman of Substance (Emma Harte Saga #1)         2.0   
997                                          Christine         3.0   
998  The Magicians' Guild (Black Magician Trilogy  #1)         2.0   
999                           The Plot Against America         2.0   

     avg_raiting  
0       4.500000  
1       4.500000  
2       4.250000  
3       4.000000  
4       4.285714  
..           ...  
995     4.230769  
996     5.000000  
997     3.428571  
998     3.500000  
999     3.000000  

[1000 rows x 3 columns]

**Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).**

In [6]:
query3 = """SELECT publisher, COUNT(book_id) as n_books 
            FROM books LEFt JOIN publishers ON books.publisher_id = publishers.publisher_id
            WHERE num_pages > 50
            GROUP BY publisher
            ORDER BY n_books DESC
            LIMIT 1
            
            """
pd.io.sql.read_sql(query3, con = engine)

publisher  n_books
0  Penguin Books       42

Penguin Books publisher is the most successfull publisher that released the biggest amount of books with more than 50 pages.

**Identify the author with the highest average book rating (look only at books with at least 50 ratings).**

In [7]:
query4 = """SELECT author, ROUND(AVG(avg_rating), 2) AS avg_rating
FROM (SELECT b.book_id, b.author_id, AVG(r.rating) AS avg_rating
  FROM books AS b 
  LEFT JOIN ratings AS r ON b.book_id = r.book_id 
  GROUP BY b.book_id 
  HAVING COUNT(r.rating_id) >= 50
  ORDER BY AVG(r.rating) DESC) AS sub
  LEFT JOIN authors AS a on sub.author_id = a.author_id
  GROUP BY author
  LIMIT 1;
"""
pd.io.sql.read_sql(query4, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.28

J.K.Rowling is one of the most rated authors with the highest average rating.

**Find the average number of text reviews among users who rated more than 50 books.**

In [8]:
query5 = """SELECT ROUND(AVG(n_reviews), 2) AS avg_reviews
            FROM
            (
            SELECT COUNT(rvw.review_id) AS n_reviews
                  FROM 
                  (SELECT username
                  FROM ratings 
                  GROUP BY username
                  HAVING COUNT(rating_id) > 50) AS usrs
            LEFT JOIN reviews AS rvw ON usrs.username = rvw.username
            GROUP BY usrs.username) AS s
            """
pd.io.sql.read_sql(query5, con = engine)

avg_reviews
0        24.33

Active users (those, who rated more than 50 books) on average write 24 text reviews. 

In our research we concluded the next:
1. There were 819 books published after January 1, 2000.
2. The number of user reviews and the average rating for each book.
3. Penguin Books publisher is the most successfull publisher that released the biggest amount of books with more than 50 pages.
4. J.K.Rowling is one of the most rated authors with the highest average rating.
5. Active users (those, who rated more than 50 books) on average write 24 text reviews.